hibernate-entitymanager/src/main/java/org/hibernate/jpa/criteria/expression/SearchedCaseExpression.java

```diff
@@ -74,13 +75,16 @@ public SearchedCaseExpression(
 	}
 
 	public Case<R> when(Expression<Boolean> condition, Expression<? extends R> result) {
-		WhenClause whenClause = new WhenClause( condition, result );
+		WhenClause whenClause = new WhenClause(
+				condition,
+				new CastFunction( criteriaBuilder(), result.getJavaType(), (ExpressionImpl) result )
```

Comment:
Let's add some comments here about why we are wrapping this in a cast.  It is needed for the HQL parser, but that is not apparent/obvious.